File: suite/Cpstak/Cpstak.rs
fn cps_tak(x: i64, y: i64, z: i64, k: Box<dyn FnOnce(i64) -> i64>) -> i64 {
    if x <= y {
        k(z)
    } else {
        cps_tak(
            x - 1,
            y,
            z,
            Box::new(move |v1| {
                cps_tak(
                    y - 1,
                    z,
                    x,
                    Box::new(move |v2| {
                        cps_tak(z - 1, x, y, Box::new(move |v3| cps_tak(v1, v2, v3, k)))
                    }),
                )
            }),
        )
    }
}

fn tak(x: i64, y: i64, z: i64) -> i64 {
    cps_tak(x, y, z, Box::new(|a| a))
}

fn main_loop(iters: u64, x: i64, y: i64, z: i64) {
    let mut res = tak(x, y, z);
    for _ in 1..iters {
        res = tak(x, y, z);
    }
    println!("{}", res);
}

fn main() {
    let mut args = std::env::args();
    args.next();
    let iters = args
        .next()
        .expect("Missing Argument iterations")
        .parse::<u64>()
        .expect("Iterations must be a number");
    let x = args
        .next()
        .expect("Missing Argument x")
        .parse::<i64>()
        .expect("x must be a number");
    let y = args
        .next()
        .expect("Missing Argument y")
        .parse::<i64>()
        .expect("y must be a number");
    let z = args
        .next()
        .expect("Missing Argument z")
        .parse::<i64>()
        .expect("z must be a number");

    main_loop(iters, x, y, z)
}

File: suite/Cpstak/Cpstak.ml
let rec cps_tak x y z k = 
  if x<=y then k(z) 
  else 
    cps_tak (x-1) y z (fun v1 -> 
    cps_tak (y-1) z x (fun v2 -> 
    cps_tak (z-1) x y (fun v3 -> 
    cps_tak v1 v2 v3 k)))

let tak x y z = cps_tak x y z (fun a -> a)

let rec main_loop iters x y z = 
  let res = tak x y z in
  if iters = 1 then 
    print_endline (string_of_int res)
  else main_loop (iters-1) x y z

let main = 
  let iters = int_of_string Sys.argv.(1) in 
  let x = int_of_string Sys.argv.(2) in 
  let y = int_of_string Sys.argv.(3) in 
  let z = int_of_string Sys.argv.(4) in 
  main_loop iters x y z
